syz-fuzzer/fuzzer.go

```diff
@@ -283,60 +224,73 @@ func main() {
 		runTest(target, manager, *flagName, config.Executor)
 		return
 	}
-
-	needPoll := make(chan struct{}, 1)
-	needPoll <- struct{}{}
-	fuzzer := &Fuzzer{
-		name:                     *flagName,
-		outputType:               outputType,
-		config:                   config,
-		execOpts:                 execOpts,
-		workQueue:                newWorkQueue(*flagProcs, needPoll),
-		needPoll:                 needPoll,
-		manager:                  manager,
-		target:                   target,
-		timeouts:                 timeouts,
-		faultInjectionEnabled:    r.CheckResult.Features[host.FeatureFault].Enabled,
-		comparisonTracingEnabled: r.CheckResult.Features[host.FeatureComparisons].Enabled,
-		corpusHashes:             make(map[hash.Sig]struct{}),
-		checkResult:              r.CheckResult,
-		fetchRawCover:            *flagRawCover,
-		noMutate:                 r.NoMutateCalls,
-		stats:                    make([]uint64, StatCount),
-		// Queue no more than ~3 new inputs / proc.
-		parallelNewInputs: make(chan struct{}, int64(3**flagProcs)),
-		resetAccState:     *flagResetAccState,
-	}
-	gateCallback := fuzzer.useBugFrames(r, *flagProcs)
-	fuzzer.gate = ipc.NewGate(gateSize, gateCallback)
-
-	for needCandidates, more := true, true; more; needCandidates = false {
-		more = fuzzer.poll(needCandidates, nil)
-		// This loop lead to "no output" in qemu emulation, tell manager we are not dead.
-		log.Logf(0, "fetching corpus: %v, signal %v/%v (executing program)",
-			len(fuzzer.corpus), len(fuzzer.corpusSignal), len(fuzzer.maxSignal))
-	}
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
 	calls := make(map[*prog.Syscall]bool)
 	for _, id := range r.CheckResult.EnabledCalls[sandbox] {
 		calls[target.Syscalls[id]] = true
 	}
-	fuzzer.choiceTable = target.BuildChoiceTable(fuzzer.corpus, calls)
-
+	fuzzerObj := fuzzer.NewFuzzer(context.Background(), &fuzzer.Config{
+		Coverage:       config.Flags&ipc.FlagSignal > 0,
+		FaultInjection: r.CheckResult.Features[host.FeatureFault].Enabled,
+		Comparisons:    r.CheckResult.Features[host.FeatureComparisons].Enabled,
+		Collide:        execOpts.Flags&ipc.FlagThreaded > 0,
+		EnabledCalls:   calls,
+		NoMutateCalls:  r.NoMutateCalls,
+		LeakChecking:   r.CheckResult.Features[host.FeatureLeak].Enabled,
+		FetchRawCover:  *flagRawCover,
+		NeedCandidates: make(chan struct{}),
```

Comment:
> Otherwise the fuzzer goroutine will block for the whole duration of RPC call.

It shouldn't -- I didn't use a blocking channel write operation in `Fuzzer.NextInput()`. If syz-fuzzer is blocked on a select, it should accept it and start a poll, otherwise we're doing a poll anyway.

Reply:
Then notifications can be lost. Non-blocking send on async channel will be lost if nobody is waiting right now.

Reply:
Moved the channel creation to `Fuzzer` and set capacity to 1.